Prevent BearerTokenAuthenticationPolicy to cause UnobservedTaskException

miroslavpokorny opened the pull request:
- Removed rethrowing of exceptions
- Rethrow cause that result headerValueTcs.Task could be never "touched" and it will cause UnobservedTaskException later (for example when GC is performed)
- Because immediately after catch blocks there is a mechanism for returning a result of headerValueTcs.Task we can let rethrow exception at that place. It will prevent possible UnobservedTaskException

# All SDK Contribution checklist:

This checklist is used to make sure that common guidelines for a pull request are followed.
- [x] **Please open PR in `Draft` mode if it is:**
	- Work in progress or not intended to be merged.
	- Encountering multiple pipeline failures and working on fixes.
- [x] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md).**
- [x] **The pull request does not introduce [breaking changes](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md).**

### [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#general-guidelines)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#testing-guidelines)
- [ ] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#sdk-generation-guidelines)
- [ ] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code. (Track 2 only)
- [ ] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK. Please double check nuget.org current release version.

## Additional management plane SDK specific contribution checklist: 
Note: Only applies to `Microsoft.Azure.Management.[RP]` or `Azure.ResourceManager.[RP]`
 
- [ ] Include updated [management metadata](https://github.com/Azure/azure-sdk-for-net/tree/master/eng/mgmt/mgmtmetadata).
- [ ] Update AzureRP.props to add/remove version info to maintain up to date API versions.

### Management plane SDK Troubleshooting
- If this is very first SDK for a services and you are adding new service folders directly under /SDK, please add `new service` label and/or contact assigned reviewer.
- If the check fails at the `Verify Code Generation` step, please ensure:
	- Do not modify any code in generated folders.
	- Do not selectively include/remove generated files in the PR.
	- Do use `generate.ps1/cmd` to generate this PR instead of calling `autorest` directly.
	Please pay attention to the @microsoft.csharp version output after running generate.ps1. If it is lower than current released version (2.3.82), please run it again as it should pull down the latest version,

### Old outstanding PR cleanup
 Please note:
	If PRs (including draft) has been out for more than 60 days and there are no responses from our query or followups, they will be closed to maintain a concise list for our reviewers.

miroslavpokorny committed:
Prevent BearerTokenAuthenticationPolicy to cause UnobservedTaskException

- Removed rethrowing of exceptions
- Rethrow cause that result headerValueTcs.Task could be never "touched" and it will cause UnobservedTaskException later (for example when GC is performed)
- Because immediately after catch blocks there is a mechanism for returning a result of headerValueTcs.Task we can let rethrow exception at that place. It will prevent possible UnobservedTaskException

jsquire commented:
Hi @miroslavpokorny.  Thank you for your contribution and interest in helping to improve the Azure SDK experience.  I've added some tags and routed to the team member best able to offer feedback.

christothes commented:
Thanks @miroslavpokorny! Can you reproduce the `UnobservedTaskException` deterministically in a test with the `throw` still in place temporarily so that this can be validated?

miroslavpokorny committed:
✅ Added test case for UnobservedTaskException when acquiring of token fails

- UnobservedTaskException should not be raised after GC is performed
- Because test is dependent on behavior of GC it can be flaky on some environments

miroslavpokorny commented:
Hi @christothes I have added test case which on previous implementation fails because of UnobservedTaskException. Test is based on GC (because problem occurs during garbage collecting) so IMO it can be flaky in some cases/ENVs...

christothes reviewed (approved):
Looks good!
